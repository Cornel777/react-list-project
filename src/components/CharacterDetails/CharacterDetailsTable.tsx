import { TableContainer, Paper, Table, TableHead, TableBody } from "@mui/material";
import Homeworld from "./components/Homeworld/Homeworld";
import { StyledTableRow, StyledTableCell } from "./components/styled/styledTable";
import { FunctionComponent } from "react";
import { Character } from "../../types/Character";

export interface CharacterDetailsTableProps {
  data: Character
}

const CharacterDetailsTable: FunctionComponent<CharacterDetailsTableProps> = ({ data }) => {
  return (
    <TableContainer component={Paper} style={{ maxWidth: '400px' }}>
        <Table size="medium" aria-label="Character traits table">
          <TableHead>
            <StyledTableRow>
              <StyledTableCell>Character trait</StyledTableCell>
              <StyledTableCell>Value</StyledTableCell>
            </StyledTableRow>
          </TableHead>
          <TableBody>
            <StyledTableRow>
              <StyledTableCell>Hair colour</StyledTableCell>
              <StyledTableCell>{data.hair_color}</StyledTableCell>
            </StyledTableRow>
            <StyledTableRow>
              <StyledTableCell>Eye colour</StyledTableCell>
              <StyledTableCell>{data.eye_color}</StyledTableCell>
            </StyledTableRow>
            <StyledTableRow>
              <StyledTableCell>Gender</StyledTableCell>
              <StyledTableCell>{data.gender}</StyledTableCell>
            </StyledTableRow>
            <StyledTableRow>
              <StyledTableCell>Homeworld</StyledTableCell>
              <StyledTableCell>
                <Homeworld homeworldUrl={data.homeworld} />
              </StyledTableCell>
            </StyledTableRow>
          </TableBody>
        </Table>
      </TableContainer>
  )
}

export default CharacterDetailsTable;