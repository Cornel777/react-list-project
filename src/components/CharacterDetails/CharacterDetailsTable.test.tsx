import { describe, it, expect, vi, beforeEach } from 'vitest';
import { render, screen, within } from '@testing-library/react';
import CharacterDetailsTable from './CharacterDetailsTable';

vi.mock('./components/Homeworld/Homeworld', () => ({
  default: () => <div data-testid="mock-homeworld">Tatooine</div>
}));

const mockCharacter = {
  name: 'Luke Skywalker',
  hair_color: 'blond',
  eye_color: 'blue',
  gender: 'male' as const,
  homeworld: 'https://swapi.dev/api/planets/1/',
  height: '172',
  mass: '77',
  birth_year: '19BBY',
  films: [],
  species: [],
  vehicles: [],
  starships: [],
  created: '',
  edited: '',
  skin_color: '',
  url: ''
};

describe('CharacterDetailsTable', () => {

  beforeEach(() => {
    vi.resetAllMocks();
  });

  it('matches the snapshot', () => {
    const { asFragment } = render(<CharacterDetailsTable data={mockCharacter} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders all character traits', () => {
    render(<CharacterDetailsTable data={mockCharacter} />);
    
    const rows = screen.getAllByRole('row');
    expect(rows).toHaveLength(5);

    const traits = ['Hair colour', 'Eye colour', 'Gender', 'Homeworld'];
    traits.forEach(trait => {
      expect(screen.getByText(trait)).toBeInTheDocument();
    });
  });

  it('displays the correct character data', () => {
    render(<CharacterDetailsTable data={mockCharacter} />);
    
    const table = screen.getByRole('table');
    const rows = within(table).getAllByRole('row');

    expect(within(rows[1]).getByText('blond')).toBeInTheDocument();
    expect(within(rows[2]).getByText('blue')).toBeInTheDocument();
    expect(within(rows[3]).getByText('male')).toBeInTheDocument();
    expect(within(rows[4]).getByText('Tatooine')).toBeInTheDocument();
  });

  it('renders the Homeworld component', () => {
    render(<CharacterDetailsTable data={mockCharacter} />);
    const homeworld = screen.getByTestId('mock-homeworld');
    expect(homeworld).toHaveTextContent('Tatooine');
  });

  it('renders table headers', () => {
    render(<CharacterDetailsTable data={mockCharacter} />);
    const table = screen.getByRole('table');
    const headers = within(table).getAllByRole('columnheader');

    expect(headers).toHaveLength(2);
    expect(headers[0]).toHaveTextContent('Character trait');
    expect(headers[1]).toHaveTextContent('Value');
  });

  });