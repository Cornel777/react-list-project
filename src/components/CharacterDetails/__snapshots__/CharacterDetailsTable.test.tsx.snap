// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CharacterDetailsTable > matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiTableContainer-root css-y2ff7i-MuiPaper-root-MuiTableContainer-root"
    style="--Paper-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12); max-width: 400px;"
  >
    <table
      aria-label="Character traits table"
      class="MuiTable-root css-1xwxv7r-MuiTable-root"
    >
      <thead
        class="MuiTableHead-root css-1a7iywq-MuiTableHead-root"
      >
        <tr
          class="MuiTableRow-root MuiTableRow-head css-17e8nb1-MuiTableRow-root"
        >
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-hktbzt-MuiTableCell-root"
            scope="col"
          >
            Character trait
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-hktbzt-MuiTableCell-root"
            scope="col"
          >
            Value
          </th>
        </tr>
      </thead>
      <tbody
        class="MuiTableBody-root css-gmh7jj-MuiTableBody-root"
      >
        <tr
          class="MuiTableRow-root css-17e8nb1-MuiTableRow-root"
        >
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1l8b5g1-MuiTableCell-root"
          >
            Hair colour
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1l8b5g1-MuiTableCell-root"
          >
            blond
          </td>
        </tr>
        <tr
          class="MuiTableRow-root css-17e8nb1-MuiTableRow-root"
        >
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1l8b5g1-MuiTableCell-root"
          >
            Eye colour
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1l8b5g1-MuiTableCell-root"
          >
            blue
          </td>
        </tr>
        <tr
          class="MuiTableRow-root css-17e8nb1-MuiTableRow-root"
        >
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1l8b5g1-MuiTableCell-root"
          >
            Gender
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1l8b5g1-MuiTableCell-root"
          >
            male
          </td>
        </tr>
        <tr
          class="MuiTableRow-root css-17e8nb1-MuiTableRow-root"
        >
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1l8b5g1-MuiTableCell-root"
          >
            Homeworld
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1l8b5g1-MuiTableCell-root"
          >
            <div
              data-testid="mock-homeworld"
            >
              Tatooine
            </div>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</DocumentFragment>
`;
